xml_converter/src/xml_converter.cpp

```diff
@@ -130,50 +149,58 @@ void write_protobuf_file(string proto_filepath, map<string, Category>* marker_ca
+                    auto category = proto_categories.find(lowercase(split_categories[split_categories.size() - 1]));
+                    if (category == proto_categories.end()) {
+                        cout << "Unknown MarkerCategory " << proto_icon.category().name() << endl;
+                    }
+                    else {
+                        // This loop adds the name of the category and all of it's parents
+                        string name;
+                        for (unsigned int j = 0; j < split_categories.size(); j++) {
+                            name += split_categories[j];
+                            category_includes.insert(name);
+                            name += ".";
+                        }
                     }
+                    output_message.add_icon()->CopyFrom(proto_icon);
                 }
-                output_message.add_icon()->CopyFrom(proto_pois.icon(i));
             }
-        }
-        for (int i = 0; i < proto_pois.trail_size(); i++) {
-            if (proto_pois.trail(i).map_id() == map_id) {
-                vector<string> split_categories = split(proto_pois.trail(i).category().name(), ".");
-                auto category = proto_categories.find(lowercase(split_categories[split_categories.size() - 1]));
-                if (category == proto_categories.end()) {
-                    cout << "Unknown MarkerCategory " << proto_pois.trail(i).category().name() << endl;
-                }
-                else {
-                    // This loop adds the name of the category and all of it's parents
-                    string name;
-                    for (unsigned int j = 0; j < split_categories.size(); j++) {
-                        name += split_categories[j];
-                        category_includes.insert(name);
-                        name += ".";
+            else if (parsed_poi->classname() == "Trail") {
+                Trail* trail = dynamic_cast<Trail*>(parsed_poi);
+                if (trail->map_id == map_id) {
+                    waypoint::Trail proto_trail = trail->as_protobuf();
+                    vector<string> split_categories = split(proto_trail.category().name(), ".");
+                    auto category = proto_categories.find(lowercase(split_categories[split_categories.size() - 1]));
+                    if (category == proto_categories.end()) {
+                        cout << "Unknown MarkerCategory " << proto_trail.category().name() << endl;
+                    }
+                    else {
+                        // This loop adds the name of the category and all of it's parents
+                        string name;
+                        for (unsigned int j = 0; j < split_categories.size(); j++) {
+                            name += split_categories[j];
+                            category_includes.insert(name);
+                            name += ".";
+                        }
```

Comment:
This is a lot of duplicated logic, maybe abstract this to a function that takes in a `map_id` integer and a `category` string?